finesse/hardware/st10_controller.py

```diff
@@ -0,0 +1,305 @@
+        **serial_kwargs: Any,
+    ):
+        """Create a new ST10Controller with the specified serial device properties.
+
+        Args:
+            port: Serial port name
+            baudrate: Serial port baudrate
+            timeout: How long to wait for read operations (seconds)
+            serial_args: Extra arguments to Serial constructor
+            serial_kwargs: Extra keyword arguments to Serial constructor
+        """
+        if "write_timeout" not in serial_kwargs:
+            serial_kwargs["write_timeout"] = timeout
+
+        serial = Serial(port, baudrate, *serial_args, timeout=timeout, **serial_kwargs)
+        return ST10Controller(serial)
+
+    def __del__(self) -> None:
+        """Leave mirror facing downwards when finished.
+
+        This prevents dust accumulating.
+        """
+        try:
+            self.move_to("nadir")
+        except Exception as e:
+            logging.error(f"Failed to reset mirror to downward position: {e}")
+
+    def _check_device_id(self) -> None:
+        """Check that the ID is the correct one for an ST10.
+
+        Raises:
+            SerialException: Error communicating with device
+            SerialTimeoutException: Timed out waiting for response from device
+            ST10ControllerError: The device ID is not for an ST10
+        """
+        self._write("MV")
+        if self._read() != "107F024":
+            raise ST10ControllerError("Device ID indicates this is not an ST10")
+
+    def _get_input_status(self, index: int) -> bool:
+        """Read the value of the device's input status.
+
+        The input status is a boolean array represented as zeros and ones. I don't know
+        what it actually corresponds to on the device, but it is used in a couple of
+        places in the old program.
+        """
+        input_status = self._request_value("IS")
+        return input_status[index] == "1"
+
+    @property
+    def steps_per_rotation(self) -> int:
+        """Get the number of steps that correspond to a full rotation."""
+        return self.STEPS_PER_ROTATION
+
+    def home(self) -> None:
+        """Return the stepper motor to its home position.
+
+        The device's internal counter is also reset to zero.
+
+        Raises:
+            SerialException: Error communicating with device
+            SerialTimeoutException: Timed out waiting for response from device
+            ST10ControllerError: Malformed message received from device
+        """
+        # If the third (boolean) value of the input status array is set, then move the
+        # motor first. I don't know what the input status actually means, but this is
+        # how it was done in the old program, so I'm copying it here.
+        if self._get_input_status(3):
+            self.relative_move(-5000)
+
+        # Send home command; leaves mirror facing upwards
+        self._write_check("SH6H")
+
+        # Turn mirror so it's facing down
+        self.relative_move(-30130)
```

Comment:
This is highly confusing. How is it possible that to move the mirror 180 degrees (from facing upwards to facing downwards) you need to move anything other than `STEPS_PER_ROTATION/2`? I think there is some sort of inconsistency on what steps mean - or a more detailed explanation is missing.

Reply:
Good point -- I hadn't noticed this. The total number of steps per rotation *should* be 50,800, based on how the old program converts between steps and angles. I've tried it and the motor seems to move by about the correct amount (e.g. when you move to the hot black body it's pointing in the right direction), so I think this value is probably correct. I suspect that it's simply that the motor case isn't aligned with things like the hot and cold black bodies, so that's why it's rotated by an extra ~35°. In any case, I should at least probably just say that we're doing what the old program did in the comment.

@jonemurray Does this make sense based on what you know about how the device was put together?